Show message and reply retrieval progress

zekehuntergreen opened the pull request:
# Description
- Currently after a metadata sync completes, if there are messages and replies to download from the server, the client shows a message "Retrieving messages" for 2.5 seconds before adding a job per message / reply to a queue.
- Due to the Guardian's volume of sources, messages, and replies, processing all message and reply download jobs after a metadata sync has completed can take a long time (around 20 minutes). 
- We would like to provide an easy way for the user to track download progress.
   - Note: as messages are downloaded, the user can see the text being populated (i.e.`<Message not yet available>` being replaced by the decrypted message text) allowing the user to approximate overall download progress.

- This change adds message to the left side of the top nav bar showing the number of message and reply download jobs in the queue when the queue is non-empty, updating every time a job is queued or dequeued (e.g. "Retrieving 23 messages").
- #1604 Is this same change into `freedomofpress:main`

## Example

https://user-images.githubusercontent.com/17057932/205993457-551796fa-f1b1-43de-813f-fc303445cc59.mov


# Test Plan

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations

zekehuntergreen committed:
`ApiJobQueue` uses `RunnablePriorityQueue`

- add main_queue_updated signal to ApiJobQueue
- RunnablePriorityQueue inherits from PriorityQueue, takes an optional pyqtSignal
- PriorityQueue emits signal with queue length when message or reply
  download is pushed / popped

zekehuntergreen committed:
controller handles main queue updated signal by updating status message

zekehuntergreen committed:
Test message retrieval progress message

- assert main queue updated signal emitted when message and reply
  download jobs added to the queue
- assert signal not emitted when other job types added to the queue
- assert retrieval message displayed when signal emitted

philmcmahon commented on securedrop_client/logic.py:
```diff
@@ -413,6 +415,15 @@ def __init__(  # type: ignore [no-untyped-def]
+    def on_main_queue_updated(self, num_items: int) -> None:
+        if num_items > 0:
+            status_text = ngettext(
+                "Retrieving 1 message", "Retrieving {message_count} messages", num_items
```
I don't know what a .pot file is but could this bit make use of those `msgid_plural` and `msgid` fields?

zekehuntergreen replied:
This is a gettext "portable object template" file. There's a [pre-commit hook](https://github.com/guardian/securedrop-client/blob/36d426179f99c68b9f679113042a135663d0e37a/.githooks/pre-commit#L4) which extracts strings that need to be localized (those passed to `gettext` or `ngettext`) from .py files and adds them to the .pot file.
A translator would then come along and add translations, which are stored in a separate .po file for each target language. More info here: https://babel.pocoo.org/en/latest/messages.html

philmcmahon commented on securedrop_client/queue.py:
```diff
@@ -28,6 +28,35 @@
+
+    def _get_num_message_or_reply_download_jobs(self) -> int:
+        message_and_reply_download_jobs = list(
+            filter(lambda job: type(job[1]) in (MessageDownloadJob, ReplyDownloadJob), self.queue)
```
might be worth adding a comment to explain why you are doing `job[1]`? It's not clear to me here what the type of self.queue is

philmcmahon reviewed (approved):
Great work 👍🏻 

![progress gif](https://media1.giphy.com/media/fZDXPWad3Rhx0H0gDp/giphy.gif?cid=6c09b9525f8c3c03ebdf82bab9741629b21d52d119131546&rid=giphy.gif&ct=s)

gonzalo-bulnes committed:
Upgrade dependency: certifi to 2022.12.07

Co-Authored-By: Kevin O'Gorman <kog@freedom.press>

gonzalo-bulnes committed:
Ignore SQLAlchemy vulnerabitity without released mitigation

The vulnerability is fixed in sqlalchemy==2.0.0b1, and the most recent
version in the 2.x series is currently 2.0.0b4.

legoktm committed:
Merge pull request #1610 from freedomofpress/temporarily-ignore-security-vulnerability

Ignore SQLAlchemy vulnerabitity without released mitigation

eaon committed:
Merge pull request #1609 from freedomofpress/upgrade-dependencies

Upgrade dependency: certifi

zekehuntergreen commented:
Closing this pull request as the same change has been merged into `freedomofpress:main` in https://github.com/freedomofpress/securedrop-client/pull/1604.
